File: haskell/app/AOC2018/Day01.hs
module AOC2018.Day01
  ( part1,
    part2,
  )
where

import Data.Set qualified as S
import Data.Text qualified as T
import Text.Parsec qualified as P
import Util.ParseUtils (parseAoCInput)

part1 :: T.Text -> Int
part1 input = sum $ parseFrequencies input

part2 :: T.Text -> Int
part2 input = findRepeatedFrequency S.empty 0 0 $ parseFrequencies input

findRepeatedFrequency :: (Num t, Ord t) => S.Set t -> t -> Int -> [t] -> t
findRepeatedFrequency seenFrequencies prevFrequency idx frequencies =
  if currentFrequency `S.member` seenFrequencies
    then currentFrequency
    else findRepeatedFrequency newSeenFrequencies currentFrequency newIdx frequencies
  where
    newSeenFrequencies = S.insert currentFrequency seenFrequencies
    newIdx = if (idx + 1) == length frequencies then 0 else idx + 1
    currentFrequency = prevFrequency + (frequencies !! idx)

parseFrequencies :: T.Text -> [Int]
parseFrequencies input = parseAoCInput input frequenciesParser "frequenciesParser"
  where
    frequencyParser = do
      sign <- P.choice $ P.try <$> [P.char '+', P.char '-']
      num <- read <$> P.many1 P.digit
      return $ if sign == '+' then num else -num
    frequenciesParser = P.many1 $ frequencyParser <* P.optional P.newline

File: python/aoc2018/Day01.py
from random import randint

with open("input.txt", 'r') as infile:
  data = infile.readlines()

# with open("input-test1.txt", 'r') as infile:
#   data = infile.readlines()

data = [int(x) for x in data]

sum = 0
for s in data:
  sum += int(s)
print(sum)

trackingSums = {}
sum = 0
idx = 0
while True:
  if sum in trackingSums.keys():
    trackingSums[sum] += 1
  else:
    trackingSums[sum] = 1
  if trackingSums[sum] > 1:
    break
  sum += data[idx]
  idx = (idx + 1) % len(data)

print(sum)
